### PR TITLE
Added a formula evaluator to replace TFormula calls

### PR DESCRIPTION
**Pull request by Dr15Jones**
Created a reco::FormulaEvaluator class which can parse a subset of
the TFormula expressions. The parsing and function evaluation are
done using stateless classes so are thread safe and extremely thread
efficient. The formula parser is shared by all reco::FormulaEvaluators
to save memory and since the parser is stateless it can efficiently
be called by multiple threads simultaneously.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
My intentions is to try this in ```SimpleJetCorrector``` to avoid the serious serialization problem it is having with ```TFormula``` when we use ROOT 6.02. The serialization problem is gone when we use ROOT 6.04 but we still pay a large startup penalty in time and memory with the use of cling.

File: CommonTools/Utils/src/formulaEvaluatorBase.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+// forward declarations
+
+namespace reco {
+  namespace formula {
+    class EvaluatorBase
+    {
+      
+    public:
+      enum class Precidence {
```
Comment (slava77) [inline CommonTools/Utils/src/formulaEvaluatorBase.h]: is it Precedence or my poor English?

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/11516	61a724b
- new code added to parse TFormula syntax and call relevant functions. Unit tests completion suggest the code integrity is fine.
- jenkins tests pass: static analysis is clean, no differences as expected

File: CommonTools/Utils/src/formulaEvaluatorBase.h (hunk 1 above)
Comment (Dr15Jones) [inline CommonTools/Utils/src/formulaEvaluatorBase.h]: Just because I'm a native speaker doesn't mean my English is any good :)

**Comment by Dr15Jones** [comment]
With this change added on top of the changes to ROOT and the pull request #11469 I am getting significantly improved performance on an Tier 0 like job when using 8 threads. I'm working on getting the precise measurements now but the short job I ran for VTune showed no major stalls or slowdowns due to unwanted synchronization.

**Comment by lgray** [comment]
@Dr15Jones I think I will start anew from something like this (or a hybrid of what already exists) for the StringCutObjectParser.

**Comment by lgray** [comment]
@Dr15Jones Wouldn't it be possible to cover N input args (even zero) using variadic templates? Can poke at that if needed.

**Comment by Dr15Jones** [comment]
From a threading stand point the fix I did to cache the function pointer created by cling pretty much solved the StringCutObjectParser efficiency problem.

**Comment by Dr15Jones** [comment]
The reason I could get away with doing a hand parser and not use boost spirit is everything was a double and the syntax was much simpler.

**Comment by lgray** [comment]
@Dr15Jones Hmm, should I continue with that project then? boost::spirit isn't the memory hog, it's really clang, which I think we will keep around any way due to simply reading our root data.

**Comment by lgray** [comment]
@Dr15Jones Probably a better use of my time is completely getting rid of TMVA (where possible) for BDTs.

**Comment by Dr15Jones** [comment]
Removing the use of TMVA is probably better since the cut parser inefficiency seems to be under control.

**Pull request merged**